File: programmers/프로세스.py
"""
코딩테스트 연습 > 스택/큐
"""


def solution(priorities, location):
    task = []

    for i in range(len(priorities)):
        task.append((i, priorities[i]))
    priorities.sort()

    count = 1
    while len(task) > 0:
        current = task[0]
        if current[1] >= priorities[-1]:
            if current[0] == location:
                return count
            task = task[1:]
            priorities.pop()
            count += 1
        else:
            task = task[1:]
            task.append(current)

File: programmers/프로세스.kt
import java.util.*

class Solution {
    fun solution(priorities: IntArray, location: Int): Int {
        var deque = ArrayDeque<Pair<Int, Int>>()
        var p = priorities.mapIndexed{ i, v ->
            Pair<Int, Int>(v, i)
        }
        deque.addAll(p)

        var count = 0
        while (deque.size > 0) {
            var top = deque.pollFirst()

            if (deque.filter{ it.first > top.first }.size > 0) {
                deque.addLast(top)
            } else {
                count = count + 1
                if (top.second == location) {
                    return count
                }
            }
        }

        return -1
    }
}
